### PR TITLE
PM - Get rid of some weak functions

### PR DESCRIPTION
**Pull request by ceolin**
This pr changes how pm policies and notifications are done. Instead of using weak functions that can be override by the application this pr defines new types and APIs for that. I kept the same behavior and name conventions for now, the idea is to do incremental steps. Currently I took slightly different approaches between the notification and policy API. For the policy, the API allows application to extend the basic policy, that is a little bit different from what we have being done so, any feedback is welcome here.

The next step, is to define a better name convention for power management, organize headers, power states, ... (ensure that points agreed in https://github.com/zephyrproject-rtos/zephyr/pull/27315 are addressed)

### CODE REVIEW

File: doc/reference/power_management/index.rst [hunk 1]
```diff
@@ -153,12 +153,41 @@ Kconfig option) less than or equal to the scheduled system idle time duration.
...
 .. code-block:: c
 
-   enum power_states sys_pm_policy_next_state(int32_t ticks);
+   struct example_pm_policy {
+         /* base pm_policy should be declared in the beginning of the struct */
+         PM_POLICY_BASE;
+         /* Custom fields */
+   };
+
+Finally, a policiy type must be defined using
```
Comment (tbursztyka) [inline doc/reference/power_management/index.rst:171]: typo

File: include/power/pm_policy.h [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+	     ({ __ASSERT(iterator <= _CONCAT(_##struct_type, _list_end), \
+			 "unexpected list end location");		\
+		     iterator < _CONCAT(_##struct_type, _list_end); }); \
+	     iterator = (struct struct_type *)				\
+		     ((uint8_t *)iterator + iterator->len))
+
+/**
+ * @brief Function to create device PM list
+ */
+void sys_pm_create_device_list(void);
```
Comment (tbursztyka) [inline include/power/pm_policy.h:83]: hopefully this will disappear someday (list will be build time generated etc...), and it won't be necessary to mention any "PM device list" like below.

File: include/power/pm_policy.h [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+static inline enum power_states sys_pm_policy_next_state(int32_t ticks)
+{
+	extern struct pm_policy *sys_pm_policy;
+
+	if (sys_pm_policy == NULL) {
+		return -EINVAL;
+	}
+
+	if (sys_pm_policy->next_state_get == NULL) {
+		/* Should it be -ENOSYS ? */
```
Comment (tbursztyka) [inline include/power/pm_policy.h:117]: -ENOTSUP seems fine to me

File: include/power/pm_policy.h [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+static inline bool sys_pm_policy_low_power_devices(enum power_states state)
+{
+	extern struct pm_policy *sys_pm_policy;
+
+	if (sys_pm_policy == NULL) {
+		return -EINVAL;
+	}
+
+	if (sys_pm_policy->low_power_devices_get == NULL) {
+		/* Should it be -ENOSYS ? */
```
Comment (tbursztyka) [inline include/power/pm_policy.h:137]: same

File: include/power/power.h [hunk 5]
```diff
@@ -64,6 +64,43 @@ extern unsigned char sys_pm_idle_exit_notify;
...
+ * @brief Statically define a power management notifier.
+ *
+ * Define a power managment notifier.
+ *
+ * @param name Name of the pm_notifier variable.
+ * @param entry_fn Function to be called when platform enters a power state
+ * @param exit_fn  Function to be called when platform exits a power state
+ */
+#define PM_NOTIFIER_DEFINE(name, entry_fn, exit_fn)		\
+	static Z_STRUCT_SECTION_ITERABLE(pm_notifier, name) =	\
```
Comment (pabigot) [inline include/power/power.h:96]: We should get some clarity on when objects are aggregated into an array by the linker, and when they are not.

If `pm_notifier` were a kernel object the aggregation would make sense.  Otherwise this API doesn't seem ideal, as it assumes that all notifiers are identified at build time and are enabled unconditionally at boot.  This forces dependencies: a module may need to set things up before its notifier is ready to be called, which won't have been done if the system does any suspension before the module is initialized (which may not happen at boot).

I think a register/unregister API that adds/removes a notifier is more flexible and has only a small cost (a link node in the notifier structure).

File: include/power/pm_policy.h (hunk 2 above)
Comment (pabigot) [reply include/power/pm_policy.h:83]: The list of devices participating in power management cannot be static: devices may be enabled or disabled, or even (in the future) instantiated at runtime.

File: include/power/pm_policy.h [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+	     iterator = (struct struct_type *)				\
+		     ((uint8_t *)iterator + iterator->len))
+
+/**
+ * @brief Function to create device PM list
+ */
+void sys_pm_create_device_list(void);
+
+/**
+ * @brief Function to suspend the devices in PM device list
```
Comment (pabigot) [inline include/power/pm_policy.h:86]: All these need to clearly define what "suspend", "low_power", "force_suspend" mean.

File: include/power/pm_policy.h [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+ * @param policy_name The policy name to be used. @see PM_POLICY_DEFINE
+ *
+ * @return false if the policy is not available, true otherwise.
+ */
+static inline bool sys_pm_policy_set(const char *policy_name)
+{
+	extern struct pm_policy *sys_pm_policy;
+
+	PM_POLICY_FOREACH(pm_policy, policy) {
+		if (policy->name != NULL &&
```
Comment (pabigot) [inline include/power/pm_policy.h:156]: is there a valid use for a policy that doesn't have a name, if they're identified by name?

File: include/power/pm_policy.h [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+}
+
+/**
+ * @brief Function to set the system power management policy
+ *
+ * @param policy_name The policy name to be used. @see PM_POLICY_DEFINE
+ *
+ * @return false if the policy is not available, true otherwise.
+ */
+static inline bool sys_pm_policy_set(const char *policy_name)
```
Comment (pabigot) [inline include/power/pm_policy.h:151]: What will be invoking this function?  Is it reasonable to do it by name, rather than passing a reference to the policy structure the application wants to use?

File: include/power/pm_policy.h [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+{
+	extern struct pm_policy *sys_pm_policy;
```
Comment (pabigot) [inline include/power/pm_policy.h:153]: why does this have to be global?  It should probably be `z_sys_pm_policy` if it is, to indicate nobody except system PM is to touch it.

File: doc/reference/power_management/index.rst (hunk 1 above)
Comment (pabigot) [reply doc/reference/power_management/index.rst:171]: also next line

File: include/power/pm_policy.h [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+ * Different policies can have their own struct type according with
+ * their necessity, the only requirements is the struct contains this
+ * macro in the beginning of the declaration and be statically defined
+ * using @c PM_POLICY_DEFINE.
+ */
+#define PM_POLICY_BASE \
+	struct { \
+		const char *name; \
+		const size_t len; \
+		enum power_states (*next_state_get)(struct pm_policy *policy,\
```
Comment (pabigot) [inline include/power/pm_policy.h:30]: I think the first argument should be `const struct pm_policy *`.  There should be no reason to have mutable data in the policy structure: if it needs state, it should store it somewhere else.

**Review by pabigot — CHANGES_REQUESTED** [review]
We should get some clarity on when objects are aggregated into an array by the linker, and when they are not.

If `pm_notifier` were a kernel object the aggregation would make sense.  Otherwise this API doesn't seem ideal, as it assumes that all notifiers are identified at build time and are enabled unconditionally at boot.  This forces dependencies: a module may need to set things up before its notifier is ready to be called, which won't have been done if the system does any suspension before the module is initialized (which may not happen at boot).

I think a register/unregister API that adds/removes a notifier is more flexible and has only a small cost (a link node in the notifier structure).

This may or may not also apply to policies, but I'd like to see a more clear description of what it would mean to dynamically add/remove/change the system power policy.  The system state under one policy would be unlikely to support the needs of another policy, and without seeing the events that produced it the new policy may not be able to accurately reconstruct the system state.

File: include/power/pm_policy.h (hunk 2 above)
Comment (tbursztyka) [reply include/power/pm_policy.h:83]: disabled (or non initialized) would be silently skipped from the whole dev list. Seems simpler that way than creating a list at runtime each time you need to do something around pm.

File: include/power/pm_policy.h (hunk 3 above)
Comment (tbursztyka) [reply include/power/pm_policy.h:117]: ok

File: include/power/pm_policy.h (hunk 2 above)
Comment (pabigot) [reply include/power/pm_policy.h:83]: Agreed: having PM capture and maintain a list of devices doesn't make sense, unless the cost of skipping over devices that don't support PM becomes so excessive that the cost of maintaining a separate list is worth doing.  I'd start assuming it's not worth it.

Comment (ceolin) [reply include/power/pm_policy.h:83]: agreed, this is not part of the policy though. I didn't touch it yet, just move it around.

File: include/power/pm_policy.h (hunk 3 above)
Comment (ceolin) [reply include/power/pm_policy.h:117]: Initially I had used ENOSYS, but then in checkpatch.pl we have "ENOSYS means 'invalid syscall nr' and nothing else\n", that is not the case here and I saw some places using ENOTSUP. I think ENOSYS is better, but I tried to be more consistent with the rest of the code base.

File: include/power/pm_policy.h (hunk 7 above)
Comment (ceolin) [reply include/power/pm_policy.h:156]: Nope, it is just a sanitycheck

File: include/power/pm_policy.h (hunk 8 above)
Comment (ceolin) [reply include/power/pm_policy.h:151]: This is just more convenient for the application, instead of having to do:
```
extern struct pm_policy *dummy_policy;
sys_pm_policy_set(dummy_policy);
```

File: include/power/pm_policy.h (hunk 9 above)
Comment (ceolin) [reply include/power/pm_policy.h:153]: yep, that makes sense.

File: include/power/pm_policy.h (hunk 10 above)
Comment (ceolin) [reply include/power/pm_policy.h:30]: I was in doubt about just adding a pointer that can be used by the implementation to store its context or allow extending the type. I can change it, in the end it will not increase the struct size because I can get rid of `len` and it will simplify some things.

File: include/power/pm_policy.h (hunk 6 above)
Comment (ceolin) [reply include/power/pm_policy.h:86]: Intend to work on it when changing the power states. This API much likely will need to change, this pr is just adding the foundation for that.

File: include/power/power.h (hunk 5 above)
Comment (ceolin) [reply include/power/power.h:96]: They can be initialized with NULL pointer callbacks in this case, we can go with the linked list approach, I have nothing against that. What I want here is confirm whether or not makes sense allow multiple notifiers.
Btw, make it kernel objects may make sense if applications want to use it from user mode and I have the feeling that this will be a valid case.

File: include/power/pm_policy.h (hunk 8 above)
Comment (pabigot) [reply include/power/pm_policy.h:151]: Why would the policy be extern?  Isn't the app providing the policy, so it'd not be an external global symbol, but a static local one?  If there are default policies available to reference they should be declared in a header somewhere.

File: include/power/pm_policy.h (hunk 3 above)
Comment (tbursztyka) [reply include/power/pm_policy.h:117]: The issue with checkpatch is that it's entirely Linux centric, which generates troubles as this one sometimes.

File: include/power/pm_policy.h [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+					      enum power_states state); \
+	}
+
+/**
+ * @brief PM Policy definition
+ *
+ * Store information about a power management policy.
+ */
+struct pm_policy {
+	PM_POLICY_BASE;
```
Comment (pabigot) [inline include/power/pm_policy.h:42]: While we're talking: I'd really rather not see the common parts of the structure provided by a define.  It should be a standalone structure, which will be the first member of  a specific policy structure, or accessed by `CONTAINER_OF`.  Field declarations in a macro are really hard to read.

Comment (ceolin) [reply include/power/pm_policy.h]: I just changed to something simpler ... Just use a basic type wouldn't work as I wanted, since it was supposed to be statically allocated by the `PM_POLICY_DEFINE` macro. I simplified it and now it is not possible to extends the type.

File: include/power/pm_policy.h [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+extern "C" {
+#endif
+
+/**
+ * @brief PM Policy definition
+ *
+ * Store information about a power management policy.
+ */
+struct pm_policy {
+	enum power_states (*next_state_get)(struct pm_policy *policy,
```
Comment (pabigot) [inline include/power/pm_policy.h:24]: The signature has changed to add an error, but as a parameter.  For consistency with other API this should return an `int` success/error code, and store the next state as an output parameter.

File: doc/reference/power_management/index.rst [hunk 13]
```diff
@@ -153,12 +153,23 @@ Kconfig option) less than or equal to the scheduled system idle time duration.
 Application
 -----------
 
-The power management policy is defined by the application which has to implement
-the following function.
+The power management policy is define by the application implementing
+the methods declared in :code:`struct pm_policy` and declared using
+:code:`PM_POLICIY_DEFINE()` macro.
```
Comment (pabigot) [inline doc/reference/power_management/index.rst:158]: s/POLICIY/POLICY/

File: include/power/power.h [hunk 14]
```diff
@@ -64,6 +64,26 @@ extern unsigned char sys_pm_idle_exit_notify;
  * @{
  */
 
+/**
+ * Power managment notifier struct
+ *
+ * This struct contains callbacks that are called when the target enters and
+ * exits power states.
```
Comment (pabigot) [inline include/power/power.h:71]: We need documentation here, and/or at the register/unregister functions, that explains when the set of notifiers can be changed.  The code iterates over the list without any protection for the list changing as a result of these functions being invoked.

**Review by pabigot — COMMENTED** [review]
A few additional comments based on current status.  Some of my previous ones may still be relevant.

**Comment by ceolin** [comment]
> A few additional comments based on current status. Some of my previous ones may still be relevant.

Thanks for reviewing it. I'll check them again, but I believe they were either addressed or became invalid with latest changes

File: include/power/pm_policy.h (hunk 8 above)
Comment (ceolin) [reply include/power/pm_policy.h]: The idea is that the symbol is implicitly created by the macro, and can be in a different object file, not necessarily in the same  file that the application will configure the policy, this is the same behavior of other similar macros.

Though as we had discussed in a previous meeting, there is no support for multiple policies  anymore, so this is not relevant anymore.

File: include/power/power.h (hunk 14 above)
Comment (ceolin) [reply include/power/power.h:71]: thinking about it, that is trick, the list is accessed from ISR as well, what makes mutex nonviable. If I use a semaphore I can't wait if its unavailable. In this case if the semaphore is not available I can't access the list and consequently not send notifications. The other problem is that suspend is not only called from the idle thread, it can be directly called from the application. Ideas ?

Comment (pabigot) [reply include/power/power.h:72]: I don't see how this has been resolved, so unresolving it.

AFAICT these are invoked from `_sys_suspend` and `_sys_resume`.  Are those from interrupt context?  Do the notification callbacks have to be invoked from that context, or can/should they be pushed to a work queue?

**Review by pabigot — CHANGES_REQUESTED** [review]
Needs to be rebased now that things have been renamed.

File: include/power/power.h (hunk 14 above)
Comment (ceolin) [reply include/power/power.h:72]: @pabigot Forgot to document it. I just protected the list of notifiers with a k_spin_lock to avoid race condition when accessing these structs. Will document that these callbacks can be called from an interruption context, at this point I don't know if that is problem, so I'll keep this behavior since that is how is currently done.

In the future if needed, I change it to a workqueue. This won't change the API anyway.

File: include/power/pm_policy.h [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+
+/**
+ * @brief Function to suspend the devices in PM device list
+ */
+int pm_suspend_devices(void);
+
+/**
+ * @brief Function to put the devices in PM device list in low power state
+ */
+int pm_low_power_devices(void);
```
Comment (pabigot) [inline include/power/pm_policy.h:59]: It would be much easier to review this if the renaming of the original version of this file were done in a separate commit from the changes to the file content.  It's not obvious what's new, and what's just refactoring existing API.

File: include/power/pm_policy.h [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+ * @brief Statically define and register a power management policy
+ *
+ * @param policy_name The name of this policy
+ * @param policy_next_state_get Callback to get the next PM state
+ * @param policy_low_power_devices_get Callback to determine whether to put
+ *        or not devices in low power state, given the system PM state
+ */
+#define PM_POLICY_DEFINE(policy_name, policy_next_state_get,		\
+			 policy_low_power_devices_get)			\
+	const struct pm_policy policy_name = {				\
```
Comment (pabigot) [inline include/power/pm_policy.h:40]: This should probably be `static` unless there's a reason to make it globally visible.

File: include/power/pm_policy.h [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+ * @return 0 in case of success, a negative value otherwise
+ */
+static inline int pm_policy_next_state(int32_t ticks,
+				       struct pm_state_info *info)
+{
+	extern const struct pm_policy *z_sys_pm_policy;
+
+	if ((z_sys_pm_policy) == NULL ||
+	    (z_sys_pm_policy->next_state_get == NULL)) {
+		return -ENOSYS;
```
Comment (pabigot) [inline include/power/pm_policy.h:83]: Should be `-ENOTSUP`.  See #23727.

File: include/power/pm_policy.h [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+
+/**
+ * @brief Function to resume the devices in PM device list
+ */
+void pm_resume_devices(void);
+
+/**
+ * @brief Function to get the next PM state based on the ticks
+ *
+ * @return 0 in case of success, a negative value otherwise
```
Comment (pabigot) [inline include/power/pm_policy.h:74]: Please document the known values.  It's fine to have a fallback for future-proofing.  E.g.

```
 * @return 
 *   * 0 in case of success
 *   * -ENOTSUP if there is no policy, or the police dpes not support a next-state capability;
 *   * other negative error codes indicating specific failure modes
```

File: include/power/pm_policy.h [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+ * @return 1 if devices should be put in low power state, 0 if not, otherwise
+ *         a negative value with the error.
+ */
+static inline int pm_policy_low_power_devices(enum pm_state state)
+{
+	extern const struct pm_policy *z_sys_pm_policy;
+
+	if ((z_sys_pm_policy == NULL) ||
+	    (z_sys_pm_policy->low_power_devices_get == NULL)) {
+		return -ENOSYS;
```
Comment (pabigot) [inline include/power/pm_policy.h:103]: `-ENOTSUP`.

**Review by pabigot — CHANGES_REQUESTED** [review]
I'd appreciate a split of the move from the content change.  I don't see much significant here, so if it's mostly moving things around the review could be shorter.  There does appear to be some new capability that could be done another way.
